patches/server/1060-Implement-preliminary-API-for-working-with-new-mob-A.patch

```diff
@@ -0,0 +1,861 @@
++import com.google.common.collect.ImmutableList;
++import com.google.common.collect.ImmutableSet;
++import com.google.common.collect.Maps;
++import com.google.common.collect.Sets;
++import com.google.common.collect.ImmutableList.Builder;
++import com.mojang.datafixers.util.Pair;
++import com.mojang.logging.LogUtils;
++import com.mojang.serialization.Codec;
++import com.mojang.serialization.DataResult;
++import com.mojang.serialization.Dynamic;
++import com.mojang.serialization.DynamicOps;
++import com.mojang.serialization.MapCodec;
++import com.mojang.serialization.MapLike;
++import com.mojang.serialization.RecordBuilder;
++import it.unimi.dsi.fastutil.objects.ObjectArrayList;
++import java.util.Collection;
++import java.util.List;
++import java.util.Map;
++import java.util.Optional;
++import java.util.Set;
++import java.util.Map.Entry;
++import java.util.function.Supplier;
++import java.util.stream.Stream;
++import javax.annotation.Nullable;
++import net.minecraft.core.registries.BuiltInRegistries;
++import net.minecraft.server.level.ServerLevel;
++import net.minecraft.util.VisibleForDebug;
++import net.minecraft.world.entity.LivingEntity;
++import net.minecraft.world.entity.ai.behavior.Behavior;
++import net.minecraft.world.entity.ai.behavior.BehaviorControl;
++import net.minecraft.world.entity.ai.memory.ExpirableValue;
++import net.minecraft.world.entity.ai.memory.MemoryModuleType;
++import net.minecraft.world.entity.ai.memory.MemoryStatus;
++import net.minecraft.world.entity.ai.sensing.Sensor;
++import net.minecraft.world.entity.ai.sensing.SensorType;
++import net.minecraft.world.entity.schedule.Activity;
++import net.minecraft.world.entity.schedule.Schedule;
++import org.apache.commons.lang3.mutable.MutableObject;
++import org.slf4j.Logger;
++
++public class Brain<E extends LivingEntity> {
++    static final Logger LOGGER = LogUtils.getLogger();
++    private final Supplier<Codec<Brain<E>>> codec;
++    private static final int SCHEDULE_UPDATE_DELAY = 20;
++    private final Map<MemoryModuleType<?>, Optional<? extends ExpirableValue<?>>> memories = Maps.newHashMap();
++    private final Map<SensorType<? extends Sensor<? super E>>, Sensor<? super E>> sensors = Maps.newLinkedHashMap();
++    private final Map<Integer, Map<Activity, Set<BehaviorControl<? super E>>>> availableBehaviorsByPriority = Maps.newTreeMap();
++    private Schedule schedule = Schedule.EMPTY;
++    private final Map<Activity, Set<Pair<MemoryModuleType<?>, MemoryStatus>>> activityRequirements = Maps.newHashMap();
++    private final Map<Activity, Set<MemoryModuleType<?>>> activityMemoriesToEraseWhenStopped = Maps.newHashMap();
++    private Set<Activity> coreActivities = Sets.newHashSet();
++    private final Set<Activity> activeActivities = Sets.newHashSet();
++    private Activity defaultActivity = Activity.IDLE;
++    private long lastScheduleUpdate = -9999L;
++
++    public static <E extends LivingEntity> Brain.Provider<E> provider(
```

Comment:
Import it how you should. See contribution.md on how to handle NMS changes.

Reply:
done